In [108]:
import pandas as pd
data = pd.read_csv("./train.csv")
data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

In [109]:
test_data = pd.read_csv("./test.csv")
test_data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave   NaN      Reg   
1     1462          20       RL         81.0    14267   Pave   NaN      IR1   
2     1463          60       RL         74.0    13830   Pave   NaN      IR1   
3     1464          60       RL         78.0     9978   Pave   NaN      IR1   
4     1465         120       RL         43.0     5005   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1454  2915         160       RM         21.0     1936   Pave   NaN      Reg   
1455  2916         160       RM         21.0     1894   Pave   NaN      Reg   
1456  2917          20       RL        160.0    20000   Pave   NaN      Reg   
1457  2918          85       RL         62.0    10441   Pave   NaN      Reg   
1458  2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence  \
0            Lvl    AllPub  ...         120        0    NaN  MnPrv   
1            Lvl    AllPub  ...           0        0    NaN    NaN   
2            Lvl    AllPub  ...           0        0    NaN  MnPrv   
3            Lvl    AllPub  ...           0        0    NaN    NaN   
4            HLS    AllPub  ...         144        0    NaN    NaN   
...          ...       ...  ...         ...      ...    ...    ...   
1454         Lvl    AllPub  ...           0        0    NaN    NaN   
1455         Lvl    AllPub  ...           0        0    NaN    NaN   
1456         Lvl    AllPub  ...           0        0    NaN    NaN   
1457         Lvl    AllPub  ...           0        0    NaN  MnPrv   
1458         Lvl    AllPub  ...           0        0    NaN    NaN   

     MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
0            NaN       0      6    2010        WD         Normal  
1           Gar2   12500      6    2010        WD         Normal  
2            NaN       0      3    2010        WD         Normal  
3            NaN       0      6    2010        WD         Normal  
4            NaN       0      1    2010        WD         Normal  
...          ...     ...    ...     ...       ...            ...  
1454         NaN       0      6    2006        WD         Normal  
1455         NaN       0      4    2006        WD        Abnorml  
1456         NaN       0      9    2006        WD        Abnorml  
1457        Shed     700      7    2006        WD         Normal  
1458         NaN       0     11    2006        WD         Normal  

[1459 rows x 80 columns]

In [110]:
data.shape

(1460, 81)

In [151]:
data.isnull().sum()

Id                    0
MSSubClass            0
MSZoning_0            0
MSZoning_1            0
MSZoning_2            0
                     ..
SaleCondition         0
MasVnrType_BrkFace    0
MasVnrType_Stone      0
MasVnrType_nan        0
SalePrice             0
Length: 83, dtype: int64

In [147]:
test_data.isnull().sum()

Id                    0
MSSubClass            0
MSZoning_0            0
MSZoning_1            0
MSZoning_2            0
                     ..
SaleType              1
SaleCondition         0
MasVnrType_BrkFace    0
MasVnrType_Stone      0
MasVnrType_nan        0
Length: 82, dtype: int64

In [146]:
data.dtypes

Id                     int64
MSSubClass             int64
MSZoning_0             int64
MSZoning_1             int64
MSZoning_2             int64
                       ...  
SaleCondition         object
MasVnrType_BrkFace      bool
MasVnrType_Stone        bool
MasVnrType_nan          bool
SalePrice              int64
Length: 83, dtype: object

In the PoolQC I have 1453 missing points from 1460 so i decided to remove this feature

In [114]:
data = data.drop('PoolQC', axis=1)
test_data = test_data.drop('PoolQC', axis=1)
test_data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave   NaN      Reg   
1     1462          20       RL         81.0    14267   Pave   NaN      IR1   
2     1463          60       RL         74.0    13830   Pave   NaN      IR1   
3     1464          60       RL         78.0     9978   Pave   NaN      IR1   
4     1465         120       RL         43.0     5005   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1454  2915         160       RM         21.0     1936   Pave   NaN      Reg   
1455  2916         160       RM         21.0     1894   Pave   NaN      Reg   
1456  2917          20       RL        160.0    20000   Pave   NaN      Reg   
1457  2918          85       RL         62.0    10441   Pave   NaN      Reg   
1458  2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities  ... 3SsnPorch ScreenPorch PoolArea  Fence  \
0            Lvl    AllPub  ...         0         120        0  MnPrv   
1            Lvl    AllPub  ...         0           0        0    NaN   
2            Lvl    AllPub  ...         0           0        0  MnPrv   
3            Lvl    AllPub  ...         0           0        0    NaN   
4            HLS    AllPub  ...         0         144        0    NaN   
...          ...       ...  ...       ...         ...      ...    ...   
1454         Lvl    AllPub  ...         0           0        0    NaN   
1455         Lvl    AllPub  ...         0           0        0    NaN   
1456         Lvl    AllPub  ...         0           0        0    NaN   
1457         Lvl    AllPub  ...         0           0        0  MnPrv   
1458         Lvl    AllPub  ...         0           0        0    NaN   

     MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
0            NaN       0      6    2010        WD         Normal  
1           Gar2   12500      6    2010        WD         Normal  
2            NaN       0      3    2010        WD         Normal  
3            NaN       0      6    2010        WD         Normal  
4            NaN       0      1    2010        WD         Normal  
...          ...     ...    ...     ...       ...            ...  
1454         NaN       0      6    2006        WD         Normal  
1455         NaN       0      4    2006        WD        Abnorml  
1456         NaN       0      9    2006        WD        Abnorml  
1457        Shed     700      7    2006        WD         Normal  
1458         NaN       0     11    2006        WD         Normal  

[1459 rows x 79 columns]

here we will deal with alley (1369 missing points)

In [115]:
import numpy as np
data['Alley'] = data['Alley'].fillna('nan')
test_data['Alley'] = data['Alley'].fillna('nan')
np.unique(data["Alley"].values)

array(['Grvl', 'Pave', 'nan'], dtype=object)

In [116]:
Alley_mapping = {"nan": 0, "Grvl": 1, "Pave": 2}
data["Alley"] = data["Alley"].map(Alley_mapping)
test_data["Alley"] = test_data["Alley"].map(Alley_mapping)
np.unique(test_data["Alley"])

array([0, 1, 2], dtype=int64)

dealing with LotFrontage : i decided to assign 0 to the missing points (~250 missing points)

In [117]:
data['LotFrontage'] = data['LotFrontage'].fillna(0)
test_data["LotFrontage"] = test_data['LotFrontage'].fillna(0)

dealing with MasVnrArea and MasVnrType we have 8, 872 missed points respectively

In [118]:
data[data["MasVnrArea"].isna()]["MasVnrType"]

234     NaN
529     NaN
650     NaN
936     NaN
973     NaN
977     NaN
1243    NaN
1278    NaN
Name: MasVnrType, dtype: object

In [119]:
# the missing points in MasVnrArea are None in MasVnrType so we will substitute them with 0
data['MasVnrArea'] = data['MasVnrArea'].fillna(0)
test_data["MasVnrArea"] = test_data['MasVnrArea'].fillna(0)

In [120]:
# for MasVnrType -> one-hot encoding
data['MasVnrType'] = data['MasVnrType'].fillna('nan')
test_data['MasVnrType'] = data['MasVnrType'].fillna('nan')

In [121]:
df_encoded = pd.get_dummies(data, columns=['MasVnrType'], drop_first=True)
sale = df_encoded['SalePrice']
df_encoded = df_encoded.drop('SalePrice', axis=1)
df_encoded.insert(len(df_encoded.columns), "SalePrice", sale)
data = df_encoded
test_data = pd.get_dummies(test_data, columns=['MasVnrType'], drop_first=True)
test_data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave      0      Reg   
1     1462          20       RL         81.0    14267   Pave      0      IR1   
2     1463          60       RL         74.0    13830   Pave      0      IR1   
3     1464          60       RL         78.0     9978   Pave      0      IR1   
4     1465         120       RL         43.0     5005   Pave      0      IR1   
...    ...         ...      ...          ...      ...    ...    ...      ...   
1454  2915         160       RM         21.0     1936   Pave      2      Reg   
1455  2916         160       RM         21.0     1894   Pave      0      Reg   
1456  2917          20       RL        160.0    20000   Pave      0      Reg   
1457  2918          85       RL         62.0    10441   Pave      0      Reg   
1458  2919          60       RL         74.0     9627   Pave      0      Reg   

     LandContour Utilities  ...  Fence MiscFeature MiscVal MoSold YrSold  \
0            Lvl    AllPub  ...  MnPrv         NaN       0      6   2010   
1            Lvl    AllPub  ...    NaN        Gar2   12500      6   2010   
2            Lvl    AllPub  ...  MnPrv         NaN       0      3   2010   
3            Lvl    AllPub  ...    NaN         NaN       0      6   2010   
4            HLS    AllPub  ...    NaN         NaN       0      1   2010   
...          ...       ...  ...    ...         ...     ...    ...    ...   
1454         Lvl    AllPub  ...    NaN         NaN       0      6   2006   
1455         Lvl    AllPub  ...    NaN         NaN       0      4   2006   
1456         Lvl    AllPub  ...    NaN         NaN       0      9   2006   
1457         Lvl    AllPub  ...  MnPrv        Shed     700      7   2006   
1458         Lvl    AllPub  ...    NaN         NaN       0     11   2006   

     SaleType SaleCondition  MasVnrType_BrkFace  MasVnrType_Stone  \
0          WD        Normal                True             False   
1          WD        Normal               False             False   
2          WD        Normal                True             False   
3          WD        Normal               False             False   
4          WD        Normal                True             False   
...       ...           ...                 ...               ...   
1454       WD        Normal               False             False   
1455       WD       Abnorml               False             False   
1456       WD       Abnorml               False              True   
1457       WD        Normal               False             False   
1458       WD        Normal               False             False   

      MasVnrType_nan  
0              False  
1               True  
2              False  
3               True  
4              False  
...              ...  
1454            True  
1455            True  
1456           False  
1457            True  
1458            True  

[1459 rows x 81 columns]

Deal with Bsmt_s
I did noticed that the same nombre of nan in ~ the same in all of them so i will see what its there?
so NA mean - No Basement

In [122]:
data.dtypes

Id                      int64
MSSubClass              int64
MSZoning               object
LotFrontage           float64
LotArea                 int64
                       ...   
SaleCondition          object
MasVnrType_BrkFace       bool
MasVnrType_Stone         bool
MasVnrType_nan           bool
SalePrice               int64
Length: 82, dtype: object

In [123]:
for col in data.columns:
    if not col.startswith("Bsmt"):
        continue
    if data[col].dtype == "object":
        data[col] = data[col].fillna("nan")
        test_data[col] = test_data[col].fillna("nan")
    else:
        data[col] = data[col].fillna(0)
        test_data[col] = test_data[col].fillna(0)

In [124]:
data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape  \
0        1          60       RL         65.0     8450   Pave      0      Reg   
1        2          20       RL         80.0     9600   Pave      0      Reg   
2        3          60       RL         68.0    11250   Pave      0      IR1   
3        4          70       RL         60.0     9550   Pave      0      IR1   
4        5          60       RL         84.0    14260   Pave      0      IR1   
...    ...         ...      ...          ...      ...    ...    ...      ...   
1455  1456          60       RL         62.0     7917   Pave      0      Reg   
1456  1457          20       RL         85.0    13175   Pave      0      Reg   
1457  1458          70       RL         66.0     9042   Pave      0      Reg   
1458  1459          20       RL         68.0     9717   Pave      0      Reg   
1459  1460          20       RL         75.0     9937   Pave      0      Reg   

     LandContour Utilities  ... MiscFeature MiscVal MoSold YrSold SaleType  \
0            Lvl    AllPub  ...         NaN       0      2   2008       WD   
1            Lvl    AllPub  ...         NaN       0      5   2007       WD   
2            Lvl    AllPub  ...         NaN       0      9   2008       WD   
3            Lvl    AllPub  ...         NaN       0      2   2006       WD   
4            Lvl    AllPub  ...         NaN       0     12   2008       WD   
...          ...       ...  ...         ...     ...    ...    ...      ...   
1455         Lvl    AllPub  ...         NaN       0      8   2007       WD   
1456         Lvl    AllPub  ...         NaN       0      2   2010       WD   
1457         Lvl    AllPub  ...        Shed    2500      5   2010       WD   
1458         Lvl    AllPub  ...         NaN       0      4   2010       WD   
1459         Lvl    AllPub  ...         NaN       0      6   2008       WD   

     SaleCondition MasVnrType_BrkFace  MasVnrType_Stone  MasVnrType_nan  \
0           Normal               True             False           False   
1           Normal              False             False            True   
2           Normal               True             False           False   
3          Abnorml              False             False            True   
4           Normal               True             False           False   
...            ...                ...               ...             ...   
1455        Normal              False             False            True   
1456        Normal              False              True           False   
1457        Normal              False             False            True   
1458        Normal              False             False            True   
1459        Normal              False             False            True   

      SalePrice  
0        208500  
1        181500  
2        223500  
3        140000  
4        250000  
...         ...  
1455     175000  
1456     210000  
1457     266500  
1458     142125  
1459     147500  

[1460 rows x 82 columns]

In [125]:
test_data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave      0      Reg   
1     1462          20       RL         81.0    14267   Pave      0      IR1   
2     1463          60       RL         74.0    13830   Pave      0      IR1   
3     1464          60       RL         78.0     9978   Pave      0      IR1   
4     1465         120       RL         43.0     5005   Pave      0      IR1   
...    ...         ...      ...          ...      ...    ...    ...      ...   
1454  2915         160       RM         21.0     1936   Pave      2      Reg   
1455  2916         160       RM         21.0     1894   Pave      0      Reg   
1456  2917          20       RL        160.0    20000   Pave      0      Reg   
1457  2918          85       RL         62.0    10441   Pave      0      Reg   
1458  2919          60       RL         74.0     9627   Pave      0      Reg   

     LandContour Utilities  ...  Fence MiscFeature MiscVal MoSold YrSold  \
0            Lvl    AllPub  ...  MnPrv         NaN       0      6   2010   
1            Lvl    AllPub  ...    NaN        Gar2   12500      6   2010   
2            Lvl    AllPub  ...  MnPrv         NaN       0      3   2010   
3            Lvl    AllPub  ...    NaN         NaN       0      6   2010   
4            HLS    AllPub  ...    NaN         NaN       0      1   2010   
...          ...       ...  ...    ...         ...     ...    ...    ...   
1454         Lvl    AllPub  ...    NaN         NaN       0      6   2006   
1455         Lvl    AllPub  ...    NaN         NaN       0      4   2006   
1456         Lvl    AllPub  ...    NaN         NaN       0      9   2006   
1457         Lvl    AllPub  ...  MnPrv        Shed     700      7   2006   
1458         Lvl    AllPub  ...    NaN         NaN       0     11   2006   

     SaleType SaleCondition  MasVnrType_BrkFace  MasVnrType_Stone  \
0          WD        Normal                True             False   
1          WD        Normal               False             False   
2          WD        Normal                True             False   
3          WD        Normal               False             False   
4          WD        Normal                True             False   
...       ...           ...                 ...               ...   
1454       WD        Normal               False             False   
1455       WD       Abnorml               False             False   
1456       WD       Abnorml               False              True   
1457       WD        Normal               False             False   
1458       WD        Normal               False             False   

      MasVnrType_nan  
0              False  
1               True  
2              False  
3               True  
4              False  
...              ...  
1454            True  
1455            True  
1456           False  
1457            True  
1458            True  

[1459 rows x 81 columns]

In [126]:
# here i will just get the target Column just to work comfortably.
sale = data['SalePrice']
data = data.drop('SalePrice', axis=1)

For the decoding we will use the mapping be cuz the order if meaningful

In [127]:
mapping = {
    "BsmtQual": {"Ex": 1, "Gd": 2, "TA": 3, "Fa": 4, "Po": 5, "nan": 6},
    "BsmtCond": {"Ex": 1, "Gd": 2, "TA": 3, "Fa": 4, "Po": 5, "nan": 6},
    "BsmtExposure": {"Gd": 1, "Av": 2, "Mn": 3, "No": 4, "nan": 5},
    "BsmtFinType1": {"GLQ": 1, "ALQ": 2, "BLQ": 3, "Rec": 4, "LwQ": 5, "Unf": 6, "nan": 7},
    "BsmtFinType2": {"GLQ": 1, "ALQ": 2, "BLQ": 3, "Rec": 4, "LwQ": 5, "Unf": 6, "nan": 7},
}
for col in mapping:
    data[col] = data[col].map(mapping[col])
    test_data[col] = test_data[col].map(mapping[col])

FireplaceQu : NA no FirePlace so ..

In [128]:
rating_mapping = {
    "Ex": 5,
    "Gd": 4,
    "TA": 3,
    "Fa": 2,
    "Po": 1,
    "nan": 0
}
data["FireplaceQu"] = data["FireplaceQu"].fillna("nan")
data["FireplaceQu"] = data["FireplaceQu"].map(rating_mapping)
test_data["FireplaceQu"] = test_data["FireplaceQu"].fillna("nan")
test_data["FireplaceQu"] = test_data["FireplaceQu"].map(rating_mapping)
data["FireplaceQu"]

0       0
1       3
2       3
3       4
4       3
       ..
1455    3
1456    3
1457    4
1458    0
1459    0
Name: FireplaceQu, Length: 1460, dtype: int64

Deal with Garage_s
so NA mean - No Garage

In [129]:
for col in data.columns:
    if not col.startswith("Garage"):
        continue
    # print(data[col].dtype)
    # print(data[col].isnull().sum())
    if data[col].dtype == "object":
        data[col] = data[col].fillna("nan")
        test_data[col] = test_data[col].fillna("nan")

In [130]:
mapping = {
    "GarageType": {
        "2Types": 1,
        "Attchd": 2,
        "Basment": 3,
        "BuiltIn": 4,
        "CarPort": 5,
        "Detchd": 6,
        "nan": 7
    },
    "GarageFinish": {
        "Fin": 1,
        "RFn": 2,
        "Unf": 3,
        "nan": 4
    },
    "GarageQual": {
        "Ex": 1,
        "Gd": 2,
        "TA": 3,
        "Fa": 4,
        "Po": 5,
        "nan": 6
    },
    "GarageCond": {
        "Ex": 1,
        "Gd": 2,
        "TA": 3,
        "Fa": 4,
        "Po": 5,
        "nan": 6
    }
}
for col in mapping:
    data[col] = data[col].map(mapping[col])
    test_data[col] = test_data[col].map(mapping[col])


In [131]:
data["GarageYrBlt"] = data["GarageYrBlt"].fillna(0)
test_data["GarageYrBlt"] = test_data["GarageYrBlt"].fillna(0)

In [132]:
data["MiscVal"]

0          0
1          0
2          0
3          0
4          0
        ... 
1455       0
1456       0
1457    2500
1458       0
1459       0
Name: MiscVal, Length: 1460, dtype: int64

so i decided to drop the MiscFeature cuz we can just rely on the MiscVal

In [133]:
data = data.drop('MiscFeature', axis=1)
test_data = test_data.drop('MiscFeature', axis=1)
data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape  \
0        1          60       RL         65.0     8450   Pave      0      Reg   
1        2          20       RL         80.0     9600   Pave      0      Reg   
2        3          60       RL         68.0    11250   Pave      0      IR1   
3        4          70       RL         60.0     9550   Pave      0      IR1   
4        5          60       RL         84.0    14260   Pave      0      IR1   
...    ...         ...      ...          ...      ...    ...    ...      ...   
1455  1456          60       RL         62.0     7917   Pave      0      Reg   
1456  1457          20       RL         85.0    13175   Pave      0      Reg   
1457  1458          70       RL         66.0     9042   Pave      0      Reg   
1458  1459          20       RL         68.0     9717   Pave      0      Reg   
1459  1460          20       RL         75.0     9937   Pave      0      Reg   

     LandContour Utilities  ... PoolArea  Fence MiscVal MoSold YrSold  \
0            Lvl    AllPub  ...        0    NaN       0      2   2008   
1            Lvl    AllPub  ...        0    NaN       0      5   2007   
2            Lvl    AllPub  ...        0    NaN       0      9   2008   
3            Lvl    AllPub  ...        0    NaN       0      2   2006   
4            Lvl    AllPub  ...        0    NaN       0     12   2008   
...          ...       ...  ...      ...    ...     ...    ...    ...   
1455         Lvl    AllPub  ...        0    NaN       0      8   2007   
1456         Lvl    AllPub  ...        0  MnPrv       0      2   2010   
1457         Lvl    AllPub  ...        0  GdPrv    2500      5   2010   
1458         Lvl    AllPub  ...        0    NaN       0      4   2010   
1459         Lvl    AllPub  ...        0    NaN       0      6   2008   

     SaleType SaleCondition  MasVnrType_BrkFace  MasVnrType_Stone  \
0          WD        Normal                True             False   
1          WD        Normal               False             False   
2          WD        Normal                True             False   
3          WD       Abnorml               False             False   
4          WD        Normal                True             False   
...       ...           ...                 ...               ...   
1455       WD        Normal               False             False   
1456       WD        Normal               False              True   
1457       WD        Normal               False             False   
1458       WD        Normal               False             False   
1459       WD        Normal               False             False   

      MasVnrType_nan  
0              False  
1               True  
2              False  
3               True  
4              False  
...              ...  
1455            True  
1456           False  
1457            True  
1458            True  
1459            True  

[1460 rows x 80 columns]

Fence

In [134]:
data["Fence"] = data["Fence"].fillna("NA")
test_data["Fence"] = test_data["Fence"].fillna("NA")
fence_mapping = {
        "GdPrv": 1,
        "MnPrv": 2,
        "GdWo": 3,
        "MnWw": 4,
        "NA": 5
    }
data["Fence"] = data["Fence"].map(fence_mapping)
test_data["Fence"] = test_data["Fence"].map(fence_mapping)

### add the sale

In [135]:
data.insert(len(data.columns), "SalePrice", sale)

In [136]:
data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape  \
0        1          60       RL         65.0     8450   Pave      0      Reg   
1        2          20       RL         80.0     9600   Pave      0      Reg   
2        3          60       RL         68.0    11250   Pave      0      IR1   
3        4          70       RL         60.0     9550   Pave      0      IR1   
4        5          60       RL         84.0    14260   Pave      0      IR1   
...    ...         ...      ...          ...      ...    ...    ...      ...   
1455  1456          60       RL         62.0     7917   Pave      0      Reg   
1456  1457          20       RL         85.0    13175   Pave      0      Reg   
1457  1458          70       RL         66.0     9042   Pave      0      Reg   
1458  1459          20       RL         68.0     9717   Pave      0      Reg   
1459  1460          20       RL         75.0     9937   Pave      0      Reg   

     LandContour Utilities  ... Fence MiscVal MoSold YrSold SaleType  \
0            Lvl    AllPub  ...     5       0      2   2008       WD   
1            Lvl    AllPub  ...     5       0      5   2007       WD   
2            Lvl    AllPub  ...     5       0      9   2008       WD   
3            Lvl    AllPub  ...     5       0      2   2006       WD   
4            Lvl    AllPub  ...     5       0     12   2008       WD   
...          ...       ...  ...   ...     ...    ...    ...      ...   
1455         Lvl    AllPub  ...     5       0      8   2007       WD   
1456         Lvl    AllPub  ...     2       0      2   2010       WD   
1457         Lvl    AllPub  ...     1    2500      5   2010       WD   
1458         Lvl    AllPub  ...     5       0      4   2010       WD   
1459         Lvl    AllPub  ...     5       0      6   2008       WD   

     SaleCondition MasVnrType_BrkFace  MasVnrType_Stone  MasVnrType_nan  \
0           Normal               True             False           False   
1           Normal              False             False            True   
2           Normal               True             False           False   
3          Abnorml              False             False            True   
4           Normal               True             False           False   
...            ...                ...               ...             ...   
1455        Normal              False             False            True   
1456        Normal              False              True           False   
1457        Normal              False             False            True   
1458        Normal              False             False            True   
1459        Normal              False             False            True   

      SalePrice  
0        208500  
1        181500  
2        223500  
3        140000  
4        250000  
...         ...  
1455     175000  
1456     210000  
1457     266500  
1458     142125  
1459     147500  

[1460 rows x 81 columns]

In [150]:
data = data.dropna(axis=0)

### MSZoning: Identifies the general zoning classification of the sale.
		
       A	Agriculture
       C	Commercial
       FV	Floating Village Residential
       I	Industrial
       RH	Residential High Density
       RL	Residential Low Density
       RP	Residential Low Density Park 
       RM	Residential Medium Density

In [138]:
test_data["MSZoning"] = test_data["MSZoning"].fillna("RM")

in MSZoning i choose to do the binary encoding

In [139]:
#!pip install category_encoders

In [140]:
import category_encoders as ce

encoder = ce.BinaryEncoder(cols=['MSZoning'])
test_data = encoder.fit_transform(test_data)
data = encoder.fit_transform(data)
data

Id  MSSubClass  MSZoning_0  MSZoning_1  MSZoning_2  LotFrontage  \
0        1          60           0           0           1         65.0   
1        2          20           0           0           1         80.0   
2        3          60           0           0           1         68.0   
3        4          70           0           0           1         60.0   
4        5          60           0           0           1         84.0   
...    ...         ...         ...         ...         ...          ...   
1455  1456          60           0           0           1         62.0   
1456  1457          20           0           0           1         85.0   
1457  1458          70           0           0           1         66.0   
1458  1459          20           0           0           1         68.0   
1459  1460          20           0           0           1         75.0   

      LotArea Street  Alley LotShape  ... Fence MiscVal MoSold YrSold  \
0        8450   Pave      0      Reg  ...     5       0      2   2008   
1        9600   Pave      0      Reg  ...     5       0      5   2007   
2       11250   Pave      0      IR1  ...     5       0      9   2008   
3        9550   Pave      0      IR1  ...     5       0      2   2006   
4       14260   Pave      0      IR1  ...     5       0     12   2008   
...       ...    ...    ...      ...  ...   ...     ...    ...    ...   
1455     7917   Pave      0      Reg  ...     5       0      8   2007   
1456    13175   Pave      0      Reg  ...     2       0      2   2010   
1457     9042   Pave      0      Reg  ...     1    2500      5   2010   
1458     9717   Pave      0      Reg  ...     5       0      4   2010   
1459     9937   Pave      0      Reg  ...     5       0      6   2008   

     SaleType SaleCondition MasVnrType_BrkFace MasVnrType_Stone  \
0          WD        Normal               True            False   
1          WD        Normal              False            False   
2          WD        Normal               True            False   
3          WD       Abnorml              False            False   
4          WD        Normal               True            False   
...       ...           ...                ...              ...   
1455       WD        Normal              False            False   
1456       WD        Normal              False             True   
1457       WD        Normal              False            False   
1458       WD        Normal              False            False   
1459       WD        Normal              False            False   

     MasVnrType_nan  SalePrice  
0             False     208500  
1              True     181500  
2             False     223500  
3              True     140000  
4             False     250000  
...             ...        ...  
1455           True     175000  
1456          False     210000  
1457           True     266500  
1458           True     142125  
1459           True     147500  

[1460 rows x 83 columns]

In [141]:
# "Street" we just have two categories
data["Street"] = data["Street"].map({"Grvl" : 0, "Pave" : 1})
test_data["Street"] = test_data["Street"].map({"Grvl" : 0, "Pave" : 1})
test_data

Id  MSSubClass  MSZoning_0  MSZoning_1  MSZoning_2  LotFrontage  \
0     1461          20           0           0           1         80.0   
1     1462          20           0           1           0         81.0   
2     1463          60           0           1           0         74.0   
3     1464          60           0           1           0         78.0   
4     1465         120           0           1           0         43.0   
...    ...         ...         ...         ...         ...          ...   
1454  2915         160           0           1           1         21.0   
1455  2916         160           0           1           1         21.0   
1456  2917          20           0           1           0        160.0   
1457  2918          85           0           1           0         62.0   
1458  2919          60           0           1           0         74.0   

      LotArea  Street  Alley LotShape  ... PoolArea Fence MiscVal MoSold  \
0       11622       1      0      Reg  ...        0     2       0      6   
1       14267       1      0      IR1  ...        0     5   12500      6   
2       13830       1      0      IR1  ...        0     2       0      3   
3        9978       1      0      IR1  ...        0     5       0      6   
4        5005       1      0      IR1  ...        0     5       0      1   
...       ...     ...    ...      ...  ...      ...   ...     ...    ...   
1454     1936       1      2      Reg  ...        0     5       0      6   
1455     1894       1      0      Reg  ...        0     5       0      4   
1456    20000       1      0      Reg  ...        0     5       0      9   
1457    10441       1      0      Reg  ...        0     2     700      7   
1458     9627       1      0      Reg  ...        0     5       0     11   

     YrSold SaleType SaleCondition MasVnrType_BrkFace MasVnrType_Stone  \
0      2010       WD        Normal               True            False   
1      2010       WD        Normal              False            False   
2      2010       WD        Normal               True            False   
3      2010       WD        Normal              False            False   
4      2010       WD        Normal               True            False   
...     ...      ...           ...                ...              ...   
1454   2006       WD        Normal              False            False   
1455   2006       WD       Abnorml              False            False   
1456   2006       WD       Abnorml              False             True   
1457   2006       WD        Normal              False            False   
1458   2006       WD        Normal              False            False   

      MasVnrType_nan  
0              False  
1               True  
2              False  
3               True  
4              False  
...              ...  
1454            True  
1455            True  
1456           False  
1457            True  
1458            True  

[1459 rows x 82 columns]

In [145]:
data['LotShape'] = data["LotShape"].map({"Reg" : 0, "IR1" : 1, "IR2" : 2, "IR3" : 3})
test_data['LotShape'] = test_data["LotShape"].map({"Reg" : 0, "IR1" : 1, "IR2" : 2, "IR3" : 3})
test_data

Id  MSSubClass  MSZoning_0  MSZoning_1  MSZoning_2  LotFrontage  \
0     1461          20           0           0           1         80.0   
1     1462          20           0           1           0         81.0   
2     1463          60           0           1           0         74.0   
3     1464          60           0           1           0         78.0   
4     1465         120           0           1           0         43.0   
...    ...         ...         ...         ...         ...          ...   
1454  2915         160           0           1           1         21.0   
1455  2916         160           0           1           1         21.0   
1456  2917          20           0           1           0        160.0   
1457  2918          85           0           1           0         62.0   
1458  2919          60           0           1           0         74.0   

      LotArea  Street  Alley  LotShape  ... PoolArea Fence MiscVal MoSold  \
0       11622       1      0         0  ...        0     2       0      6   
1       14267       1      0         1  ...        0     5   12500      6   
2       13830       1      0         1  ...        0     2       0      3   
3        9978       1      0         1  ...        0     5       0      6   
4        5005       1      0         1  ...        0     5       0      1   
...       ...     ...    ...       ...  ...      ...   ...     ...    ...   
1454     1936       1      2         0  ...        0     5       0      6   
1455     1894       1      0         0  ...        0     5       0      4   
1456    20000       1      0         0  ...        0     5       0      9   
1457    10441       1      0         0  ...        0     2     700      7   
1458     9627       1      0         0  ...        0     5       0     11   

     YrSold SaleType SaleCondition MasVnrType_BrkFace MasVnrType_Stone  \
0      2010       WD        Normal               True            False   
1      2010       WD        Normal              False            False   
2      2010       WD        Normal               True            False   
3      2010       WD        Normal              False            False   
4      2010       WD        Normal               True            False   
...     ...      ...           ...                ...              ...   
1454   2006       WD        Normal              False            False   
1455   2006       WD       Abnorml              False            False   
1456   2006       WD       Abnorml              False             True   
1457   2006       WD        Normal              False            False   
1458   2006       WD        Normal              False            False   

      MasVnrType_nan  
0              False  
1               True  
2              False  
3               True  
4              False  
...              ...  
1454            True  
1455            True  
1456           False  
1457            True  
1458            True  

[1459 rows x 82 columns]

test fill na

In [155]:
test_data.isnull().sum()

Id                    0
MSSubClass            0
MSZoning_0            0
MSZoning_1            0
MSZoning_2            0
                     ..
SaleType              0
SaleCondition         0
MasVnrType_BrkFace    0
MasVnrType_Stone      0
MasVnrType_nan        0
Length: 82, dtype: int64

In [154]:
test_data["Utilities"] = test_data["Utilities"].fillna("ELO")
test_data["Exterior1st"] = test_data["Exterior1st"].fillna("Other")
test_data["Exterior2nd"] = test_data["Exterior2nd"].fillna("Other")
test_data["TotalBsmtSF"] = test_data["TotalBsmtSF"].fillna(0)
test_data["KitchenQual"] = test_data["KitchenQual"].fillna("TA")
test_data["Functional"] = test_data["Functional"].fillna("Mod")
test_data["GarageCars"] = test_data["GarageCars"].fillna(0)
test_data["GarageArea"] = test_data["GarageArea"].fillna(0)
test_data["SaleType"] = test_data["SaleType"].fillna("Oth")

In [165]:
data.dtypes

Id                  int64
MSSubClass          int64
MSZoning_0          int64
MSZoning_1          int64
MSZoning_2          int64
                    ...  
LotConfig_Inside     bool
Electrical_FuseF     bool
Electrical_FuseP     bool
Electrical_Mix       bool
Electrical_SBrkr     bool
Length: 123, dtype: object

In [157]:
# get/add the Sales
sale = df_encoded['SalePrice']
data = data.drop('SalePrice', axis=1)

In [167]:
# data.insert(len(data.columns), "SalePrice", sale)

In [159]:
HouseStyle_mapping = {
    "1Story": 1,
    "1.5Unf": 2,
    "1.5Fin": 3,
    "2Story": 4,
    "2.5Unf": 5,
    "2.5Fin": 6,
    "SLvl": 7,
    "SFoyer": 8
}
data['HouseStyle'] = data['HouseStyle'].map(HouseStyle_mapping)
test_data['HouseStyle'] = test_data['HouseStyle'].map(HouseStyle_mapping)

In [160]:
data['LandSlope'] = data['LandSlope'].map({"Gtl" : 1, "Mod" : 2, "Sev" : 3})
test_data['LandSlope'] = test_data['LandSlope'].map({"Gtl" : 1, "Mod" : 2, "Sev" : 3})

In [158]:
test_data = pd.get_dummies(test_data, columns=['LandContour', "Utilities", "LotConfig", "Electrical"]
                           , drop_first=True)
data = pd.get_dummies(data, columns=['LandContour', "Utilities", "LotConfig", "Electrical"]
                           , drop_first=True)

In [162]:
encoder = ce.BinaryEncoder(cols=['Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'RoofStyle', 'RoofMatl',
                                 "Exterior1st", "Exterior2nd", "Foundation", "Heating", "SaleType", "SaleCondition"])
encoder.fit(data)
test_data = encoder.fit_transform(test_data)
data = encoder.fit_transform(data)
data

Id  MSSubClass  MSZoning_0  MSZoning_1  MSZoning_2  LotFrontage  \
0        1          60           0           0           1         65.0   
1        2          20           0           0           1         80.0   
2        3          60           0           0           1         68.0   
3        4          70           0           0           1         60.0   
4        5          60           0           0           1         84.0   
...    ...         ...         ...         ...         ...          ...   
1455  1456          60           0           0           1         62.0   
1456  1457          20           0           0           1         85.0   
1457  1458          70           0           0           1         66.0   
1458  1459          20           0           0           1         68.0   
1459  1460          20           0           0           1         75.0   

      LotArea  Street  Alley  LotShape  ...  LandContour_Lvl  \
0        8450       1      0         0  ...             True   
1        9600       1      0         0  ...             True   
2       11250       1      0         1  ...             True   
3        9550       1      0         1  ...             True   
4       14260       1      0         1  ...             True   
...       ...     ...    ...       ...  ...              ...   
1455     7917       1      0         0  ...             True   
1456    13175       1      0         0  ...             True   
1457     9042       1      0         0  ...             True   
1458     9717       1      0         0  ...             True   
1459     9937       1      0         0  ...             True   

      Utilities_NoSeWa  LotConfig_CulDSac  LotConfig_FR2  LotConfig_FR3  \
0                False              False          False          False   
1                False              False           True          False   
2                False              False          False          False   
3                False              False          False          False   
4                False              False           True          False   
...                ...                ...            ...            ...   
1455             False              False          False          False   
1456             False              False          False          False   
1457             False              False          False          False   
1458             False              False          False          False   
1459             False              False          False          False   

      LotConfig_Inside  Electrical_FuseF  Electrical_FuseP  Electrical_Mix  \
0                 True             False             False           False   
1                False             False             False           False   
2                 True             False             False           False   
3                False             False             False           False   
4                False             False             False           False   
...                ...               ...               ...             ...   
1455              True             False             False           False   
1456              True             False             False           False   
1457              True             False             False           False   
1458              True             False             False           False   
1459              True             False             False           False   

      Electrical_SBrkr  
0                 True  
1                 True  
2                 True  
3                 True  
4                 True  
...                ...  
1455              True  
1456              True  
1457              True  
1458             False  
1459              True  

[1459 rows x 123 columns]

In [164]:
# Mapping for ExterQual
data['ExterQual'] = data["ExterQual"].map({"Ex": 0, "Gd": 1, "TA": 2, "Fa": 3, "Po": 4})
test_data['ExterQual'] = test_data["ExterQual"].map({"Ex": 0, "Gd": 1, "TA": 2, "Fa": 3, "Po": 4})

# Mapping for ExterCond
data['ExterCond'] = data["ExterCond"].map({"Ex": 0, "Gd": 1, "TA": 2, "Fa": 3, "Po": 4})
test_data['ExterCond'] = test_data["ExterCond"].map({"Ex": 0, "Gd": 1, "TA": 2, "Fa": 3, "Po": 4})

# Mapping for HeatingQC
data['HeatingQC'] = data["HeatingQC"].map({"Ex": 0, "Gd": 1, "TA": 2, "Fa": 3, "Po": 4})
test_data['HeatingQC'] = test_data["HeatingQC"].map({"Ex": 0, "Gd": 1, "TA": 2, "Fa": 3, "Po": 4})

# Mapping for CentralAir
data['CentralAir'] = data["CentralAir"].map({"N": 0, "Y": 1})
test_data['CentralAir'] = test_data["CentralAir"].map({"N": 0, "Y": 1})

# Mapping for KitchenQual
data['KitchenQual'] = data["KitchenQual"].map({"Ex": 0, "Gd": 1, "TA": 2, "Fa": 3, "Po": 4})
test_data['KitchenQual'] = test_data["KitchenQual"].map({"Ex": 0, "Gd": 1, "TA": 2, "Fa": 3, "Po": 4})

# Mapping for PavedDrive
data['PavedDrive'] = data["PavedDrive"].map({"Y": 0, "P": 1, "N": 2})
test_data['PavedDrive'] = test_data["PavedDrive"].map({"Y": 0, "P": 1, "N": 2})

# Mapping for Functional
data['Functional'] = data["Functional"].map({
    "Typ": 0, 
    "Min1": 1, 
    "Min2": 2, 
    "Mod": 3, 
    "Maj1": 4, 
    "Maj2": 5, 
    "Sev": 6, 
    "Sal": 7
})
test_data['Functional'] = test_data["Functional"].map({
    "Typ": 0, 
    "Min1": 1, 
    "Min2": 2, 
    "Mod": 3, 
    "Maj1": 4, 
    "Maj2": 5, 
    "Sev": 6, 
    "Sal": 7
})


In [168]:
data

Id  MSSubClass  MSZoning_0  MSZoning_1  MSZoning_2  LotFrontage  \
0        1          60           0           0           1         65.0   
1        2          20           0           0           1         80.0   
2        3          60           0           0           1         68.0   
3        4          70           0           0           1         60.0   
4        5          60           0           0           1         84.0   
...    ...         ...         ...         ...         ...          ...   
1455  1456          60           0           0           1         62.0   
1456  1457          20           0           0           1         85.0   
1457  1458          70           0           0           1         66.0   
1458  1459          20           0           0           1         68.0   
1459  1460          20           0           0           1         75.0   

      LotArea  Street  Alley  LotShape  ...  Utilities_NoSeWa  \
0        8450       1      0         0  ...             False   
1        9600       1      0         0  ...             False   
2       11250       1      0         1  ...             False   
3        9550       1      0         1  ...             False   
4       14260       1      0         1  ...             False   
...       ...     ...    ...       ...  ...               ...   
1455     7917       1      0         0  ...             False   
1456    13175       1      0         0  ...             False   
1457     9042       1      0         0  ...             False   
1458     9717       1      0         0  ...             False   
1459     9937       1      0         0  ...             False   

      LotConfig_CulDSac  LotConfig_FR2  LotConfig_FR3  LotConfig_Inside  \
0                 False          False          False              True   
1                 False           True          False             False   
2                 False          False          False              True   
3                 False          False          False             False   
4                 False           True          False             False   
...                 ...            ...            ...               ...   
1455              False          False          False              True   
1456              False          False          False              True   
1457              False          False          False              True   
1458              False          False          False              True   
1459              False          False          False              True   

      Electrical_FuseF  Electrical_FuseP  Electrical_Mix  Electrical_SBrkr  \
0                False             False           False              True   
1                False             False           False              True   
2                False             False           False              True   
3                False             False           False              True   
4                False             False           False              True   
...                ...               ...             ...               ...   
1455             False             False           False              True   
1456             False             False           False              True   
1457             False             False           False              True   
1458             False             False           False             False   
1459             False             False           False              True   

      SalePrice  
0        208500  
1        181500  
2        223500  
3        140000  
4        250000  
...         ...  
1455     175000  
1456     210000  
1457     266500  
1458     142125  
1459     147500  

[1459 rows x 124 columns]

In [169]:
test_data

Id  MSSubClass  MSZoning_0  MSZoning_1  MSZoning_2  LotFrontage  \
0     1461          20           0           0           1         80.0   
1     1462          20           0           1           0         81.0   
2     1463          60           0           1           0         74.0   
3     1464          60           0           1           0         78.0   
4     1465         120           0           1           0         43.0   
...    ...         ...         ...         ...         ...          ...   
1454  2915         160           0           1           1         21.0   
1455  2916         160           0           1           1         21.0   
1456  2917          20           0           1           0        160.0   
1457  2918          85           0           1           0         62.0   
1458  2919          60           0           1           0         74.0   

      LotArea  Street  Alley  LotShape  ...  LandContour_Low  LandContour_Lvl  \
0       11622       1      0         0  ...            False             True   
1       14267       1      0         1  ...            False             True   
2       13830       1      0         1  ...            False             True   
3        9978       1      0         1  ...            False             True   
4        5005       1      0         1  ...            False            False   
...       ...     ...    ...       ...  ...              ...              ...   
1454     1936       1      2         0  ...            False             True   
1455     1894       1      0         0  ...            False             True   
1456    20000       1      0         0  ...            False             True   
1457    10441       1      0         0  ...            False             True   
1458     9627       1      0         0  ...            False             True   

      Utilities_ELO  LotConfig_CulDSac  LotConfig_FR2  LotConfig_FR3  \
0             False              False          False          False   
1             False              False          False          False   
2             False              False          False          False   
3             False              False          False          False   
4             False              False          False          False   
...             ...                ...            ...            ...   
1454          False              False          False          False   
1455          False              False          False          False   
1456          False              False          False          False   
1457          False              False          False          False   
1458          False              False          False          False   

      LotConfig_Inside  Electrical_FuseF  Electrical_FuseP  Electrical_SBrkr  
0                 True             False             False              True  
1                False             False             False              True  
2                 True             False             False              True  
3                 True             False             False              True  
4                 True             False             False              True  
...                ...               ...               ...               ...  
1454              True             False             False              True  
1455              True             False             False              True  
1456              True             False             False              True  
1457              True             False             False              True  
1458              True             False             False              True  

[1459 rows x 120 columns]

In [ ]:
# data.to_csv('processed_train_data.csv', index=False)
# test_data.to_csv('processed_test_data.csv', index=False)